mlrun/package/packagers_manager.py

```diff
@@ -0,0 +1,632 @@
+        }
+
+        # Set the instructions in the artifact's spec:
+        artifact.spec.package_instructions = package_instructions
+
+        # Collect the artifact and return:
+        self._artifacts.append(artifact)
+        return artifact
+
+    def unpack(self, data_item: DataItem, type_hint: typing.Type) -> typing.Any:
+        """
+        Unpack an object using one of the manager's packagers.
+
+        :param data_item: The data item holding the package.
+        :param type_hint: The type hint to parse the data item as.
+
+        :return: The unpacked object parsed as type hinted.
+        """
+        # Set variables to hold the manager notes and packager instructions:
+        artifact_key = None
+        packager_name = None
+        object_type = None
+        artifact_type = None
+        instructions = {}
+
+        # Try to get the notes and instructions (can be found only in artifacts but data item may be a simple path/url):
+        if (
+            data_item.artifact_url
+            and is_store_uri(url=data_item.artifact_url)
+            and parse_store_uri(data_item.artifact_url)[0]
+            in [StorePrefix.Artifact, StorePrefix.Dataset, StorePrefix.Model]
+        ):
+            # Get the artifact object in the data item:
+            artifact, _ = store_manager.get_store_artifact(url=data_item.artifact_url)
+            # Get the key from the artifact's metadata and instructions from the artifact's spec:
+            artifact_key = artifact.metadata.key
+            package_instructions = artifact.spec.package_instructions
+            # Extract the manager notes and packager instructions found:
+            if package_instructions:
+                packager_name = package_instructions.pop(
+                    self._InstructionsNotesKeys.PACKAGER_NAME, None
+                )
+                object_type = package_instructions.pop(
+                    self._InstructionsNotesKeys.OBJECT_TYPE, None
+                )
+                if object_type is not None:
+                    # Get the actual type from the stored string:
+                    object_type = self._get_type_from_name(type_name=object_type)
+                artifact_type = package_instructions.pop(
+                    self._InstructionsNotesKeys.ARTIFACT_TYPE, None
+                )
+                instructions = package_instructions.pop(
+                    self._InstructionsNotesKeys.INSTRUCTIONS, {}
+                )
+
+        # If both original packaged object type and user's type hint available, validate they are equal:
+        if object_type is not None and type_hint is not None:
+            matching_object_and_type_hint = False
+            hinted_types = [type_hint]
+            while len(hinted_types) > 0:
+                if object_type in hinted_types:
+                    matching_object_and_type_hint = True
+                    break
+                hinted_types = self._reduce_type_hints(type_hints=hinted_types)
```

Comment:
I notice that type hints in the MLRun context (i.e. for inputes/results) are handled and processed in several places - they are handled in the `run()` function, here and in other classes. I think you should concentrate the various type-hint processing functions in a single place, maybe in a class that implements all of them.

Reply:
I think that `mlrun.package` is now the place for type hints / log hints so any other place that wish to use these should take the functions from there (as at the end package will deal with all of these). I changed all use cases I know of to use `mlrun.package`.